proof/refine/ARM/Ipc_R.thy

```diff
@@ -2304,20 +2296,478 @@ crunches doIPCTransfer, possibleSwitchTo
+             apply (rule corres_split_deprecated [OF _ sts_corres])
+                apply (rule set_ep_corres)
+                apply (simp add: ep_relation_def)
+               apply (simp add: fault_rel_optionation_def)
+              apply wp+
+            apply (clarsimp simp: st_tcb_at_tcb_at valid_tcb_state_def
+                                  invs_def valid_state_def valid_pspace_def)
+           apply (clarsimp simp: invs'_def valid_state'_def valid_pspace'_def)
+           \<comment> \<open>concludes IdleEP\<close>
+          apply (simp add: ep_relation_def)
+          apply (case_tac bl; simp add: ep_relation_def)
+          apply (rule corres_guard_imp)
+            apply (rule corres_split_deprecated [OF _ sts_corres])
+               apply (rule corres_split_deprecated [OF _ tcbEPAppend_corres])
+                 apply (rule set_ep_corres)
+                 apply (simp add: ep_relation_def)
+                apply (wpsimp wp: hoare_vcg_ball_lift)+
+           apply (clarsimp simp: st_tcb_at_tcb_at valid_tcb_state_def invs_def valid_state_def valid_pspace_def valid_ep_def)
+          apply (clarsimp simp: invs'_def valid_state'_def valid_pspace'_def valid_ep'_def)
+          \<comment> \<open>concludes SendEP\<close>
+         apply (simp add: ep_relation_def)
+         apply (rename_tac list)
+         apply (rule_tac F="list \<noteq> []" in corres_req)
+          apply (simp add: valid_ep_def)
+         apply (case_tac list)
+          apply simp
+         apply (clarsimp split del: if_split)
+         \<comment> \<open>start corres logic\<close>
+         apply (rule corres_guard_imp)
+           apply (rule corres_split_deprecated [OF _ set_ep_corres])
+              apply (simp add: isReceive_def split del:if_split)
+              apply (rule corres_split_deprecated [OF _ gts_corres])
+                apply (rule stronger_corres_guard_imp)
+                apply (rule_tac
+                       F="\<exists>data pl. recv_state = Structures_A.BlockedOnReceive ep data pl"
+                       in corres_gen_asm)
+                apply (clarsimp simp: case_bool_If case_option_If if3_fold
+                            simp del: dc_simp split del: if_split cong: if_cong)
+                apply (rule corres_split_deprecated [OF _ dit_corres])
+                  apply (rule corres_split_deprecated[where r'=dc])
+                     apply (simp only: get_tcb_obj_ref_def)
+                     apply (rule corres_split_deprecated [OF _ threadget_corres[where r="(=)"]])
+                        apply (rule corres_split_deprecated[OF _ threadget_corres[where r=fault_rel_optionation]])
+                           apply (rule corres_split_deprecated[OF _ corres_if[where r=dc], where r=dc])
+                                apply (rule corres_split_deprecated [OF possibleSwitchTo_corres sts_corres])
+                                  apply clarsimp
+                                 prefer 4 \<comment> \<open>deferring Hoare triples\<close>
```

Comment:
Is there a nicer way than using `prefer 4` here? Maybe even a `rotated n` would be slightly nicer